jkind/src/jkind/solvers/yices2/Yices2Solver.java

```diff
@@ -37,13 +42,35 @@ protected String getSolverExecutable() {
 	@Override
 	public void initialize() {
 		send("(set-option :produce-models true)");
-		send("(set-logic QF_UFNIRA)");
+		send("(set-option :produce-unsat-cores true)");
+		if(linear) {
+			send("(set-logic QF_UFLIRA)");
+		} else {
+			send("(set-logic QF_UFNIRA)");
+		}
 	}
 
 	@Override
 	protected List<Symbol> getUnsatCore(List<Symbol> activationLiterals) {
-		// Yices2 does not yet support unsat-cores
-		return activationLiterals;
+		List<Symbol> unsatCore = new ArrayList<>();
+		send("(get-unsat-core)");
+		for (String s : readCore().split(" ")) {
+			if (!s.isEmpty()) {
+				unsatCore.add(new Symbol(s.substring(1)));
+			}
+		}
+		return unsatCore;		
+	}
+	
+	private String readCore() {
+		String line = "";
+		try {
+			line = fromSolver.readLine();
+			comment(getSolverName() + ": " + line);
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+		return line.substring(1, line.length() - 1);
```

Comment:
In theory we should be able to lift these two methods up to SmtLib2Solver so that we don't duplicate them in Z3 and Yices2. Can you try this and see if it works?

Reply:
slight variations:

![image](https://user-images.githubusercontent.com/2699181/57498599-08d5ab80-72a2-11e9-824c-5e1a2f2dfc28.png)